lib/puppet/provider/agent_upgrade/windows.rb

```diff
@@ -0,0 +1,85 @@
+
+Puppet::Type.type(:agent_upgrade).provide(:agent_upgrade) do
+  confine :osfamily => /windows/i
+
+  mk_resource_methods
+
+  commands :powershell =>
+             if File.exists?("#{ENV['SYSTEMROOT']}\\sysnative\\WindowsPowershell\\v1.0\\powershell.exe")
+               "#{ENV['SYSTEMROOT']}\\sysnative\\WindowsPowershell\\v1.0\\powershell.exe"
+             elsif File.exists?("#{ENV['SYSTEMROOT']}\\system32\\WindowsPowershell\\v1.0\\powershell.exe")
+               "#{ENV['SYSTEMROOT']}\\system32\\WindowsPowershell\\v1.0\\powershell.exe"
+             else
+               'powershell.exe'
+             end
+
+  def self.instances
+    [new({
+           :name => 'puppet-agent',
+           :ensure => :present,
+           :version => Facter.value(:puppetversion),
+         })
+    ]
+  end
+
+  def create
+    spawn_upgrade_process
+  end
+
+  def exists?
+    false
+  end
+
+  def source_url
+    resource[:source] ||
+      "https://downloads.puppetlabs.com/windows/#{puppet_agent_msi_filename}"
+  end
+
+  def puppet_agent_msi_filename
+    arch = resource[:arch] || Facter.value(:architecture)
+    if resource[:version] =~ /latest/i
+      return "puppet-agent-#{arch}-latest.msi"
+    end
+
+    "puppet-agent-#{resource[:version]}-#{arch}.msi"
+  end
+
+
+  def get_command
+    tempfile = new_logfile_path
+    debug "Spawning installer with log at '#{tempfile}'"
+    "cmd.exe /c \"\"#{native_path(command(:powershell))}\"" +
+      " -NoProfile -NonInteractive -NoLogo -ExecutionPolicy Bypass -Command " +
+      "\"Wait-Process #{Process.pid} -ErrorAction SilentlyContinue;msiexec /qn /i #{source_url} /l*v #{tempfile} \""
+  end
+
+  def new_logfile_path
+    native_path(Tempfile.new(['puppet-agent-installer', '.log']).path)
+  end
+
+  def spawn_upgrade_process
+    args = {
+      :command_line => get_command,
+      :creation_flags => Process::DETACHED_PROCESS,
+      :process_inherit => false,
+      :thread_inherit => true,
+      :cwd => 'C:\\'
+    }
+
+    debug "Creating wait process for #{Process.pid}"
+    upgradeProcess = Process.create(args)
+    debug ("Created background process to wait for puppet exit at PID #{upgradeProcess.process_id}")
```

Comment:
With this debug message you actually need to tell the user what you are doing and where they should follow up if the agent is not upgraded. I would create a secondary log from that process that tells them that the process was successful or not (pointing to the msi install log as well for further review), plus how they move forward with getting support for failures.